### DISCUSSION TITLE
Each message is sent twice at RTPS layer

### DISCUSSION THREAD

**Discussion by kunnalparihar** [eProsima/Fast-DDS#4198] [Support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

Hello, Thanks for the great work. I am currently using fast DDS to send data from one PC to another.
I used publisher code (Fast-DDS/examples/cpp/dds/HelloWorldExample/HelloWorldPublisher.cpp) to send 10 messages of different sizes.
At the subscriber end, I correctly receive 10 messages. But there is some issue when I check the RTPS layer.

### Current behavior

I found that at publisher side, I modified Fast-DDS/src/cpp/rtps/transport/UDPTransportInterface.cpp code to add print (cout) after sending a message.    
I added print (cout) inside the below function. I found that for each message from application layer, I see two prints at this function.    
It looks like for sending one message, rtps try sending the message to receiver twice.

UDPTransportInterface::send(const octet* send_buffer, uint32_t send_buffer_size, eProsimaUDPSocket& socket, const Locator& remote_locator, bool only_multicast_purpose, bool whitelisted,  const std::chrono::microseconds& timeout)

Can you please guide me?    

Thank you in advance.   

### Steps to reproduce

you can some cout at send function inside UDPTransportInterface.cpp code.

### Fast DDS version/commit

v2.12.0

### Platform/Architecture

Ubuntu Focal 20.04 arm64

### Transport layer

UDPv4

**Comment by Mario-DL** [comment]
Hi @kunnalparihar,

This is not really an issue. If `multicast is enabled`, for every interface (localhost is whitelisted by default) two sender resources are created, one for `unicast` and another for `multicast`. We try to send every message through all the sender resources. Downwards, in the udp sender resource, it is checked if the locator is truly a multicast address (`IPLocator::isMulticast()`). You will notice that for the user traffic sender locator, the multicast sender does not send anything, so it is sent through the unicast udp sender only. In addition, you should also check that you are not mixing discovery and user traffic messages. 

I encourage you to use `wireshark` or other packet analyzer for further inspection. I am moving this issue to a discussion.

**Reply by kunnalparihar** [reply]
@Mario-DL Hi, thanks for the explanation.
I modified my code, to publish to only one IP address but still it sends the message twice. Can you guide me on how to resolve this issue?
Thanks a lot
Below is a snippet of the code HelloWorldPublisher.cpp init function:

    pqos.transport().use_builtin_transports = false;
    int32_t kind = LOCATOR_KIND_UDPv4;
    Locator initial_peer_locator;
    initial_peer_locator.kind = kind;
    IPLocator::setIPv4(initial_peer_locator, "10.219.127.154");
    initial_peer_locator.port = 5100;
    //pqos.wire_protocol().builtin.initialPeersList.push_back(initial_peer_locator);
    pqos.wire_protocol().builtin.metatrafficUnicastLocatorList.push_back(initial_peer_locator);

    auto udp_transport = std::make_shared<UDPv4TransportDescriptor>();
    udp_transport->segment_size(2 * 1024 * 1024);
    udp_transport->non_blocking_send = false;
    //udp_transport->multicast_locator_list.clear();
    //udp_transport->remote_locator_list.push_back(initial_peer_locator);
    pqos.transport().user_transports.push_back(udp_transport);

Below is a snippet of the code HelloWorldSubscriber.cpp init function:

    pqos.transport().use_builtin_transports = false;
    int32_t kind = LOCATOR_KIND_UDPv4;
    Locator initial_peer_locator;
    initial_peer_locator.kind = kind;
    IPLocator::setIPv4(initial_peer_locator, "10.219.127.154");
    initial_peer_locator.port = 5100;
    pqos.wire_protocol().builtin.initialPeersList.push_back(initial_peer_locator);
    auto sm_transport = std::make_shared<UDPv4TransportDescriptor>();
    //sm_transport->segment_size(2 * 1024 * 1024);
    sm_transport->non_blocking_send = false;
    pqos.transport().user_transports.push_back(sm_transport);
